### DISCUSSION TITLE
Security transfer between accounts

### DISCUSSION THREAD

**Discussion by Resurrectum** [ghostfolio/ghostfolio#1899] [General]
How is it possible to transfer securities (or cash) between accounts? For instance if you want to move a pile of stocks from account `A` to `account B`, or when you just transfer some cash between accounts?

**Comment by swaggerino** [comment]
looking for a solution too (maybe has to be implemented yet?)

**Comment by dtslvr** [comment]
> For instance if you want to move a pile of stocks from account `A` to account `B`

If it’s just a one-time thing, you could export your activities, modify the `accountId` in the `JSON` file, and reimport the activities (for example into a new user account). An alternative would be to adjust the activities directly in the database.

**Comment by elnigno** [comment]
Is a new `TRANSFER` transaction type being considered?

**Reply by dtslvr** [reply]
Currently, no. I don't see a big advantage that would justify the increase in complexity by introducing a new activity type.

**Reply by joestump** [reply]
@dtslvr IRA/401k transfers are pretty common in the US. First account I tried to put into Ghostfolio was my IRA, which as 3x other IRA transfers in the last 12 months. Not entirely uncommon to see security account transfers too. 

What is the suggestion then? I have rather important activities I can't map without this kind of type.

**Reply by nipunarora** [reply]
@dtslvr - Same for me, transferring assets to another firm. What is the way we can proceed

**Reply by rafa17campos** [reply]
That's the same than: https://github.com/ghostfolio/ghostfolio/discussions/3455 isn't it? I wouldn't mind trying to implement it, although I don't work on Node so it will take some time. A good opportunity to learn :). Is it that complex? Any advice if I finally decide to go for it?

**Reply by tvirus97** [reply]
+1 from me. 

This feature will be very useful because there are many transfers and there needs be a record for each transaction. 

If we just edit activities by exporting and reimporting, we lose the details of the original transactions and can lead to confusion in future due to “false” entries, when comparing to broker/bank records. 

https://www.portfolio-performance.info has this functionality, so it is technically feasible.

**Comment by arnonm** [comment]
+1

**Comment by joeshmoe57** [comment]
Upvoted, would love to see support for this, as I just created a new account and moved the majority of my assets into it (from multiple existing accounts). Not sure how to proceed without this feature.

What are others actually doing? I'm wary to update the database and know all my data is "wrong". Best I can think of in the interim is to just map the new account to one of the old accounts, but I fear it'll get convoluted. Curious if anyone's come up with a clever workaround.

**Comment by ups7** [comment]
In-specie transfers are also very common in the UK, so it will be great to have a dedicated activity type for transfers to track everything accurately. Especially because you lose the book cost during portfolio transfers (unless the new broker supports updating it manually, which some don't). Over the years, the same holding can go through multiple transfers. Having an accurate history makes maintaining your portfolio on Ghostfolio even more valuable.

The only workaround I can think of is to create a single `SELL` activity for the old account for number of units being transferred, and then a `BUY` activity for the new account at the exact same price as the `SELL` activity. This isn't ideal but at least preserves the history of where the original transactions took place. I suspect that this will mess up some performance charts, however.

**Comment by joeshmoe57** [comment]
I ended up writing a small script to basically do what @dtslvr suggested: I re-mapped my existing activities to my new account. I added some tags in an attempt to not totally lose the data, but they don't factor in to any of the calculations, so they're optional. In my case, I transferred all assets from 2 accounts "in-kind" to a new account, so this covered that while maintaining cost basis information (the "sell" + "buy" option above would lose this
\). It may not cover all cases as-is, but should be able to be adjusted easily. Dropping here with some instructions in case it helps anyone.

0. I had all my data backed up prior to starting this
1. If you want to use tags for record keeping, create them now. Assign them to any activity so they show up in the export below
2. Download all activities from `/portfolio/activities`. This is both a backup and the ingestion data
3. (Optional) I filtered the activities to just the accounts I planned to touch: source ones I planned to migrate and destination ones I was migrating to, and downloaded those activity sets as well, just to make re-importing easier in case things went sideways
4. Make some notes for any validation you want: I took some screenshots of my holdings and summary pages so I could compare afterwards
5. Populate `migrations.json` in the format below. All names used in `migrations` must already exist and have an ID to map to (included in the export in step 2)
6. Put the export from step 2 in the same directory as `migration.json` and the script (and put the filename in `migrations.json`)
7. Run it like `python3 transform.py`
8. Import the resulting json
9. Delete whatever activities are no longer needed
10. Validate how you want via step 4

### `migrations.json`
```json
{
    "export_to_ingest": "ghostfolio-export-202603192248.json",
    "output_filename": "migrated-ghostfolio.json",
    "migrations": [
        {
            "source_name": "Old Asset Account 1",
            "destination_name": "New Asset Account",
            "destination_tag_names": [
                "Old_Account_1_Provenance",
                "Transfer_Date_2026-03-20"
            ]
        },
        {
            "source_name": "Old Asset Account 2",
            "destination_name": "New Asset Account",
            "destination_tag_names": [
                "Old_Account_2_Provenance",
                "Transfer_Date_2026-03-01"
            ]
        }
    ]
}
```
### `transform.py`
```python
import json
import copy

class MigrationAccount:

    def __init__(self, source_name, dest_tag_names, dest_name):
        self.source_name = source_name
        self.dest_tag_names = dest_tag_names
        self.dest_name = dest_name
        self.source_accountId = None
        self.dest_accountId = None
        self.dest_tag_ids = []

    @staticmethod
    def _name_to_id(name, id_name_item):
        id = None
        if id_name_item['name'] == name:
            id = id_name_item['id']
        return id

    def map_to_ids(self, accounts, tags):
        for account in accounts:
            source_id = self._name_to_id(self.source_name, account)
            if source_id:
                if self.source_accountId:
                    print('Duplicate name detected; please prevent name collisions')
                self.source_accountId = source_id
            
            dest_id = self._name_to_id(self.dest_name, account)
            if dest_id:
                if self.dest_accountId:
                    print('Duplicate name detected; please prevent name collisions')
                self.dest_accountId = dest_id
        for tag in tags:
            for tag_name in self.dest_tag_names:
                tag_id = self._name_to_id(tag_name, tag)
                if tag_id:
                    # missing name collision detection
                    self.dest_tag_ids.append(tag_id)

def _load_json(filename):
    loaded = None
    with open(filename) as _file:
        loaded = json.load(_file)
    return loaded

def load_config(config_filename='migrations.json'):
    loaded = _load_json(config_filename)
    migrations = []
    for item in loaded['migrations']:
        migration = MigrationAccount(
            source_name=item['source_name'],
            dest_name=item['destination_name'],
            dest_tag_names=item['destination_tag_names'])
        migrations.append(migration)

    return migrations, loaded['export_to_ingest'], loaded['output_filename']

def load_export(export_filename):
    return _load_json(export_filename)

migrations, export_filename, output_filename = load_config()
export = load_export(export_filename)
for migration in migrations:
    migration.map_to_ids(export['accounts'], export['tags'])
activities = export['activities']
updated_activities = []
for activity in activities:
    for migration in migrations:
        if migration.source_accountId == activity['accountId']:
            updated_activity = copy.copy(activity)
            updated_activity['accountId'] = migration.dest_accountId
            for tag in migration.dest_tag_ids:
                if tag not in updated_activity['tags']:
                    updated_activity['tags'].append(tag)
            updated_activities.append(updated_activity)

output = {}
output['activities'] = updated_activities
with open(output_filename, 'w') as output_file:
    json.dump(output, output_file)
```
